src/rust/engine/workunit_store/src/lib.rs

```diff
@@ -837,45 +843,55 @@ pub fn expect_workunit_store_handle() -> WorkunitStoreHandle {
 /// NB: Public for macro usage: use the `in_workunit!` macro.
 ///
 pub fn _start_workunit(
-  workunit_store: WorkunitStore,
+  store_handle: &mut WorkunitStoreHandle,
   name: String,
   initial_metadata: WorkunitMetadata,
-) -> (WorkunitStoreHandle, RunningWorkunit) {
-  let mut store_handle = expect_workunit_store_handle();
+) -> RunningWorkunit {
   let span_id = SpanId::new();
   let parent_id = std::mem::replace(&mut store_handle.parent_id, Some(span_id));
-  let workunit = workunit_store.start_workunit(span_id, name, parent_id, initial_metadata);
-  (store_handle, RunningWorkunit::new(workunit_store, workunit))
+  let workunit = store_handle
+    .store
+    .start_workunit(span_id, name, parent_id, initial_metadata);
+  RunningWorkunit::new(store_handle.store.clone(), Some(workunit))
 }
 
 #[macro_export]
 macro_rules! in_workunit {
-    ($workunit_store: expr, $workunit_name: expr, $workunit_metadata: expr, |$workunit: ident| async move { $( $body:tt )* $(,)? }) => (
-      {
-        let (store_handle, mut $workunit) = $crate::_start_workunit($workunit_store, $workunit_name, $workunit_metadata);
-        $crate::scope_task_workunit_store_handle(Some(store_handle), async move {
-          let result = {
-            let $workunit = &mut $workunit;
-            async move { $( $body )* }
-          }.await;
-          $workunit.complete();
-          result
-        })
-      }
-    );
-  ($workunit_store: expr, $workunit_name: expr, $workunit_metadata: expr, |$workunit: ident| $f: expr $(,)?) => (
-    {
-      let (store_handle, mut $workunit) = $crate::_start_workunit($workunit_store, $workunit_name, $workunit_metadata);
+  ($workunit_store: expr, $workunit_name: expr, $workunit_metadata: expr, |$workunit: ident| $f: expr $(,)?) => {{
+    // TODO: The workunit store argument is unused: remove in separate patch.
+    std::mem::drop($workunit_store);
+
+    use futures::future::FutureExt;
+    let mut store_handle = $crate::expect_workunit_store_handle();
+    // TODO: Move Level out of the metadata to allow skipping construction if disabled.
+    let workunit_metadata = $workunit_metadata;
+    if store_handle.store.max_level() >= workunit_metadata.level {
+      let mut $workunit =
+        $crate::_start_workunit(&mut store_handle, $workunit_name, workunit_metadata);
       $crate::scope_task_workunit_store_handle(Some(store_handle), async move {
-          let result = {
-            let $workunit = &mut $workunit;
-            $f
-          }.await;
-          $workunit.complete();
-          result
-        })
+        let result = {
+          let $workunit = &mut $workunit;
+          $f
+        }
+        .await;
+        $workunit.complete();
+        result
+      })
+      .boxed()
+    } else {
+      // TODO: This RunningWorkunit has no Workunit associated, and so counter increments will be
+      // dropped.
```

Comment:
@tdyas : This is probably a blocker: interested in your thoughts. We do frequently have counters attached to `TRACE` workunits, and as it stands, those increments will just be lost. It seems possible to fix by moving counters back to a global, or recording them on the parent instead... but the former loses our counter nesting (which we admittedly don't use), and the latter is a very likely recipe for losing metrics (á la #11548).

Reply:
I'm leaning as well toward just dropping it in favor of global counters.

My original use case for having nested counters was to be able to discern (eventually in server-side analysis) whether a work unit met some condition or not, for example, whether the remote cache was used or not, by looking at counters under a subtree of all work units. In practice though, it has only made sense to look at the counters for an entire run. (Part of that may be that Buildsense doesn't have a way to slice down to per-target counters yet, but in practice haven't had yet experienced a need for that.)

Questions:
1. Maybe put the counter increment on the nearest non-TRACE parent?
2. Do work units have a way to add structured metadata? I recall Pants [sets a cached status in the description](https://github.com/pantsbuild/pants/blob/6037e699b881011f845b5f7ad109fa8eae39cc3b/src/rust/engine/process_execution/src/remote_cache.rs#L428), but it would be much more convenient when inspecting work units on the server to set a structured label. Then there would be no need in my use case for non-global counters.

Reply:
> Maybe put the counter increment on the nearest non-TRACE parent?

Yea, we could do that. But that might mean that they end up arbitrarily far away, and different locations depending on the level.

> Do work units have a way to add structured metadata? I recall Pants sets a cached status in the description, but it would be much more convenient when inspecting work units on the server to set a structured label. Then there would be no need in my use case for non-global counters.

Yes, metadata is supported. We capture the cache status, exit code, process definition, target addresses, etc: anything captured by https://github.com/pantsbuild/pants/blob/6bf2002f257d873d633c374aef11568a38d75942/src/python/pants/engine/engine_aware.py#L29-L36

Reply:
(If we switch to global counters, I recommend doing that as prework. Pardon if that was already the plan)

Reply:
Broke out global counters over here: #14541.